### PR TITLE
feat(placeholder): add placholder component and service 

### LINKED ISSUE #22: Move all placeholder services into one shared service
dialog placeholder and modal placeholder do effectively the same thing ... they should be combined into a single service and component

### PR DESCRIPTION
**Pull request by cal-smith**
- update modal.service to use placeholder.service
- update dialog.service to use placeholder.service
- add `BaseModal` and deprecate `@Modal`
- add deprecation notices to `modal-placeholder` and `dialog-placeholder`

closes #22

### CODE REVIEW

**Comment by zvonimirfras** [comment]
It's the type, you'd still name the variable `placeholder: PlaceholderService` and keep the naming consistent.

**Comment by cal-smith** [comment]
Yeah I know, but then that's breaking the service injection convention

**Comment by zvonimirfras** [comment]
I don't understand why we need to deviate from our conventions in this case at all. What makes this so special?

**Comment by cal-smith** [comment]
It's really not ... I just wasn't satisfied with the `XService` naming ... it's `PlaceholderService` now though

File: src/dialog/dialog-placeholder.component.ts [hunk 1]
```diff
@@ -1,11 +1,10 @@
...
 	OnInit,
-	Injector,
 	ViewContainerRef,
 	ViewChild
 } from "@angular/core";
-import { DialogPlaceholderService } from "./dialog-placeholder.service";
+import { PlaceholderService } from "./../placeholder/placeholder.module";
 
 /**
  * Using a dialog (popover, tooltip, etc) with appendToBody="true" in your application
```
Comment (zvonimirfras) [inline src/dialog/dialog-placeholder.component.ts]: Maybe add a deprecation comment so we remove it for 3.0

File: src/dialog/popover/popover.directive.spec.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { By } from "@angular/platform-browser";
 import { NFormsModule } from "./../../forms/forms.module";
 import { IconModule } from "./../../icon/icon.module";
 import { StaticIconModule } from "./../../icon/static-icon.module";
-import { DialogPlaceholderService } from "./../dialog-placeholder.service";
+import { Placeholder } from "./../../placeholder/placeholder.service";
```
Comment (zvonimirfras) [inline src/dialog/popover/popover.directive.spec.ts]: Does this have to be
```suggestion
import { PlaceholderService } from "./../../placeholder/placeholder.service";
```

File: src/dialog/popover/popover.directive.spec.ts [hunk 3]
```diff
@@ -32,7 +32,7 @@ xdescribe("Popover directive", () => {
 				IconModule,
 				StaticIconModule
 			],
-			providers: [ DialogPlaceholderService ]
+			providers: [ Placeholder ]
```
Comment (zvonimirfras) [inline src/dialog/popover/popover.directive.spec.ts]:
```suggestion
			providers: [ PlaceholderService ]
```
?

File: src/modal/base-modal.class.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { Output, EventEmitter } from "@angular/core";
+
+/**
+ * Extend BaseModal in your custom modal implementations to ensure consistent close behaviour.
```
Comment (zvonimirfras) [inline src/modal/base-modal.class.ts]:
```suggestion
 * Extend `BaseModal` in your custom modal implementations to ensure consistent close behaviour.
```

File: src/modal/base-modal.class.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ *
+ * ModalService depends on the `close` event to correctly clean up the component.
```
Comment (zvonimirfras) [inline src/modal/base-modal.class.ts]:
```suggestion
 * `ModalService` depends on the `close` event to correctly clean up the component.
```

File: src/modal/modal-placeholder.component.ts [hunk 6]
```diff
@@ -45,16 +50,14 @@ export class ModalPlaceholderComponent implements OnInit {
 
 	/**
 	 * Creates an instance of `ModalPlaceholderComponent`.
-	 * @param {ModalService} modalPlaceholderService
-	 * @memberof ModalPlaceholderComponent
 	 */
-	constructor(public modalPlaceholderService: ModalPlaceholderService) {}
+	constructor(public placeholder: Placeholder) {}
```
Comment (zvonimirfras) [inline src/modal/modal-placeholder.component.ts]: Should this be `PlaceholderService`?

File: src/modal/modal.decorator.ts [hunk 7]
```diff
@@ -1,15 +1,20 @@
+import { Output, EventEmitter } from "@angular/core";
+
 /**
+ * # Deprecated - extend BaseModal instead
```
Comment (zvonimirfras) [inline src/modal/modal.decorator.ts]:
```suggestion
 * # Deprecated - extend `BaseModal` instead
```

File: src/placeholder/placeholder.component.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+} from "@angular/core";
+import { PlaceholderService } from "./placeholder.service";
+
+/**
+ * Using a modal, dialog (Tooltip, OverflowMenu), or any other component that draws out of the normal page flow
+ * in your application *requires* this component (`ibm-placeholder`).
+ * It would generally be placed near the end of your root app component template
+ * (app.component.ts or app.component.html) as:
+ *
+ * ...
```
Comment (zvonimirfras) [inline src/placeholder/placeholder.component.ts]:
```suggestion
 * ```
```
?

**Pull request merged**